src/meta-srv/src/selector/load_based.rs

```diff
@@ -12,79 +12,236 @@
-
-        // sort the datanodes according to the number of regions
-        tuples.sort_by(|a, b| a.2.cmp(&b.2));
-
-        Ok(tuples
-            .into_iter()
-            .map(|(stat_key, lease_val, _)| Peer {
-                id: stat_key.node_id,
-                addr: lease_val.node_addr,
-            })
-            .collect())
+        // 2. get stat kvs and filter out expired datanodes.
+        let stat_keys = lease_kvs.keys().map(|k| k.into()).collect();
+        let stat_kvs = filter_out_expired_datanode(
+            ctx.meta_peer_client.get_dn_stat_kvs(stat_keys).await?,
+            &lease_kvs,
+        );
+
+        // 3. make sure the regions of a table are distributed on different datanodes.
+        let stat_kvs = if let (Some(catalog), Some(schema), Some(table)) =
+            (&ctx.table, &ctx.schema, &ctx.table)
+        {
+            filter_out_datanode_by_table(stat_kvs, catalog, schema, table)
+        } else {
+            stat_kvs
+        };
+
+        // 4. check whether the data meets the SelectorOptions.
+        check(&stat_kvs, &opts)?;
+
+        // 5. compute weight array
+        let weight_array = compute_weight_array(&stat_kvs, &lease_kvs);
+
+        // 6. choose peers from weight_array
+        let mut weighted_choose = self.weighted_choose.write();
+        let selected = choose_peers(weight_array, &opts, &mut *weighted_choose);
+
+        Ok(selected)
+    }
+}
+
+pub(crate) fn filter_out_expired_datanode(
+    mut stat_kvs: HashMap<StatKey, StatValue>,
+    lease_kvs: &HashMap<LeaseKey, LeaseValue>,
+) -> HashMap<StatKey, StatValue> {
+    lease_kvs
+        .iter()
+        .filter_map(|(lease_k, _)| stat_kvs.remove_entry(&lease_k.into()))
+        .collect()
+}
+
+pub(crate) fn filter_out_datanode_by_table(
+    stat_kvs: HashMap<StatKey, StatValue>,
+    catalog: &str,
+    schema: &str,
+    table: &str,
+) -> HashMap<StatKey, StatValue> {
+    stat_kvs
+        .into_iter()
+        .filter(|(_, stat_v)| contains_table(stat_v, catalog, schema, table) != Some(true))
+        .collect()
+}
+
+pub(crate) fn check(stat_kvs: &HashMap<StatKey, StatValue>, opts: &SelectorOptions) -> Result<()> {
+    if opts.allow_duplication {
+        ensure!(
+            !stat_kvs.is_empty(),
+            error::NoEnoughAvailableDatanodeSnafu {
+                required: opts.min_required_items,
+                available: 0_usize
+            }
+        );
+    } else {
+        ensure!(
+            stat_kvs.len() >= opts.min_required_items,
+            error::NoEnoughAvailableDatanodeSnafu {
+                required: opts.min_required_items,
+                available: stat_kvs.len()
+            }
+        );
+    }
+
+    Ok(())
+}
+
+/// Compute a weight for each datanode according to stat_kvs.
+///
+/// How to compute weight? The formula is as follows.
+/// The weight of datanode = the max region number in datanodes - region number of the datanode +
+/// base weight.
+pub(crate) fn compute_weight_array(
```

Comment:
I think the calculation based on region_num is only temporary. In the future, it is more likely to be based on WRCU. Therefore, it would be best to calculate the weight using a closure or something passed in.